src/myrtlespeech/run/train.py

```diff
@@ -47,43 +47,48 @@ def fit(
 
         for stage in stages:
             is_training = stage == Stage.TRAIN
-            seq_to_seq.train(mode=is_training)
-            cb_handler.train(mode=is_training)
+            loader = train_loader if is_training else eval_loader
+            if run_stage(seq_to_seq, cb_handler, loader, is_training):
+                break
+
+    cb_handler.on_train_end()
 
-            cb_handler.on_epoch_begin()
 
-            with ExitStack() as stack:
-                if not is_training:
-                    stack.enter_context(torch.no_grad())
+def run_stage(seq_to_seq, cb_handler, loader, is_training):
+    seq_to_seq.train(mode=is_training)
+    cb_handler.train(mode=is_training)
 
-                loader = train_loader if is_training else eval_loader
-                for x, y in loader:
-                    # model
-                    x, y = cb_handler.on_batch_begin(x, y)
-                    out = seq_to_seq.model(x)
+    cb_handler.on_epoch_begin()
 
-                    # loss
-                    loss_out, loss_y = cb_handler.on_loss_begin(out, y)
-                    loss = seq_to_seq.loss(loss_out, loss_y)
-                    loss, skip_bwd = cb_handler.on_backward_begin(loss)
+    with ExitStack() as stack:
+        if not is_training:
+            stack.enter_context(torch.no_grad())
 
-                    # optim
-                    if is_training:
-                        if not skip_bwd:
-                            loss.backward()
+        for x, y in loader:
 
-                        if seq_to_seq.optim is not None:
-                            if not cb_handler.on_backward_end():
-                                seq_to_seq.optim.step()
+            x, y = cb_handler.on_batch_begin(x, y)
+            out = seq_to_seq.model(x)
 
-                            if not cb_handler.on_step_end():
-                                seq_to_seq.optim.zero_grad()
+            # loss
+            loss_out, loss_y = cb_handler.on_loss_begin(out, y)
+            loss = seq_to_seq.loss(loss_out, loss_y)
+            loss, skip_bwd = cb_handler.on_backward_begin(loss)
 
-                    if cb_handler.on_batch_end():
-                        break
+            # optim
+            if is_training:
+                if not skip_bwd:
+                    loss.backward()
 
-            if cb_handler.on_epoch_end():
+                if seq_to_seq.optim is not None:
+                    if not cb_handler.on_backward_end():
+                        seq_to_seq.optim.step()
+
+                    if not cb_handler.on_step_end():
+                        seq_to_seq.optim.zero_grad()
+            if cb_handler.on_batch_end():
                 break
 
-    cb_handler.on_train_end()
+            del x, y, out, loss, loss_out, loss_y
+
+    return cb_handler.on_epoch_end()
     # sphinx-doc-end-before
```

Comment:
The generated documentation now includes the source code from the first line of fit to here. This no longer makes sense - the documentation should be updated to reflect the separate functions.

Reply:
Done!